nxengine/libretro/libretro.cpp

```diff
@@ -90,7 +92,68 @@ void retro_get_system_info(struct retro_system_info *info)
 
 char g_dir[1024];
 
-void retro_set_controller_port_device(unsigned port, unsigned device) {}
+void retro_set_controller_port_device(unsigned port, unsigned device)
+{
+   if (port != 0) return;
+
+   memset(inputs, 0, sizeof(inputs));
+   memset(lastinputs, 0, sizeof(lastinputs));
+   memset(mappings, 0, sizeof(mappings));
+   for (unsigned i = 0; i < INPUT_COUNT; ++i)
+      mappings[i] = RETROK_DUMMY;
+
+   if (device == RETRO_DEVICE_KEYBOARD) {
+      // Use the original keybindings
+      controller_device = RETRO_DEVICE_KEYBOARD;
+      mappings[LEFTKEY]  = RETROK_LEFT;
+      mappings[RIGHTKEY] = RETROK_RIGHT;
+      mappings[UPKEY]    = RETROK_UP;
+      mappings[DOWNKEY]  = RETROK_DOWN;
+
+      mappings[JUMPKEY] = RETROK_z;
+      mappings[FIREKEY] = RETROK_x;
+
+      mappings[PREVWPNKEY] = RETROK_a;
+      mappings[NEXTWPNKEY] = RETROK_s;
+
+      mappings[INVENTORYKEY] = RETROK_q;
+      mappings[MAPSYSTEMKEY] = RETROK_w;
+
+      mappings[ESCKEY] = RETROK_ESCAPE;
+      mappings[F1KEY]  = RETROK_F1;
+      mappings[F2KEY]  = RETROK_F2;
+      mappings[F3KEY]  = RETROK_F3;
+   } else {
+      // Use a joypad model in all other cases
```

Comment:
The joypad is the default, if you set a keyboard, you get a keyboard, everything else sets up the joypad.